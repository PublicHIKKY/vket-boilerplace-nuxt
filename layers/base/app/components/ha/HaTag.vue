<template>
  <span
    class="ha-tag"
    :class="[
      `-${category}`,
      {
        '-disabled': disabled,
        '-clickable': clickable,
      },
    ]"
    :disabled="disabled"
    @click="onClick"
  >
    <slot />
  </span>
</template>

<script lang="ts">
import { ValueOf } from '#base/app/utils/types/types'
/**
 * src/components/ha/HaTag.vueにて
 * TAG_TYPEをscript setup内のローカルで宣言するとテストで落ちる。
 * その為、scriptタグを別で設置して宣言する
 * ※このscriptタグにはsetupは付けないこと。prettierで同一scriptタグに整形される為。
 */

const TAG_TYPE = {
  PRIMARY: 'primary',
  PINK: 'pink',
  OUTLINE: 'outline',
  GRAY: 'gray',
  DANGER: 'danger',
} as const
</script>

<script setup lang="ts">
type Props = {
  disabled?: boolean
  category?: ValueOf<typeof TAG_TYPE>
  clickable?: boolean
}

type Emits = {
  (emit: 'click'): void
}

const props = withDefaults(defineProps<Props>(), {
  disabled: false,
  category: TAG_TYPE.PRIMARY,
})

const emits = defineEmits<Emits>()

const onClick = () => {
  if (props.disabled) return false
  return emits('click')
}
</script>

<style lang="scss" scoped>
@use '#base/app/assets/styles/variables' as v;
@use 'sass:color';

.ha-tag {
  border-radius: 4px;
  display: inline-block;
  font-size: 12px;
  line-height: 1;
  padding: v.space(1) v.space(2);
  transition: color 0.2s ease, background-color 0.2s ease;
  width: fit-content;

  &.-clickable {
    cursor: pointer;
  }

  &.-primary {
    background-color: v.$primary-color;
    color: v.$white;
    font-weight: 600;

    &.-clickable:hover {
      background-color: v.$yellow;
      box-shadow: 0 0 6px v.$yellow;
    }

    &.-disabled {
      background-color: v.$button-disabled-color;
      color: v.$gray-3;
    }
  }

  &.-pink {
    background-color: v.$pink;
    color: v.$white;

    &.-clickable:hover {
      background-color: color.adjust(
        v.$pink,
        $saturation: 10%,
        $lightness: 10%
      );
      box-shadow: 0 0 6px v.$pink;
    }

    &.-disabled {
      background-color: v.$button-disabled-color;
      color: v.$gray-3;
    }
  }

  &.-outline {
    background-color: transparent;
    border: v.$pink 1px solid;
    color: v.$pink;

    &.-clickable:hover {
      background-color: v.$pink;
      color: v.$white;
    }

    &.-disabled {
      background-color: v.$button-disabled-color;
      color: v.$gray-3;
    }
  }

  &.-gray {
    background-color: v.$gray;
    color: v.$white;

    &.-clickable:hover {
      background-color: color.adjust(
        v.$gray,
        $saturation: 10%,
        $lightness: 10%
      );
    }

    &.-disabled {
      background-color: v.$button-disabled-color;
      color: v.$gray-3;
    }
  }

  &.-danger {
    background-color: v.$red;
    color: v.$white;

    &.-clickable:hover {
      background-color: color.adjust(v.$red, $saturation: 10%, $lightness: 10%);
    }

    &.-disabled {
      background-color: v.$button-disabled-color;
      color: v.$gray-3;
    }
  }
}
</style>
