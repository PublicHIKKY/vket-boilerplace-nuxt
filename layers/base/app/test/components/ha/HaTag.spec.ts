import { mount } from '@vue/test-utils'
import { describe, expect, test } from 'vitest'
import HaTag from '#base/app/components/ha/HaTag.vue'

describe('HaTag', () => {
  test('ref component', () => {
    expect(HaTag).toBeTruthy()
  })

  test('mount component', () => {
    const wrapper = mount(HaTag)
    expect(wrapper.getCurrentComponent()).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('slot', () => {
    const wrapper = mount(HaTag, {
      slots: {
        default: 'Slot Content',
      },
    })

    expect(wrapper.text()).toContain('Slot Content')
  })

  describe('props', () => {
    test('renders props', () => {
      const wrapper = mount(HaTag, {
        props: {
          disabled: false,
          category: 'primary',
          clickable: true,
        },
      })

      expect(wrapper.exists()).toBe(true)
    })

    test('renders with the "-tagType" when category prop is tagType', () => {
      const tagType = 'outline'

      const wrapper = mount(HaTag, {
        props: {
          category: tagType,
        },
      })

      expect(wrapper.classes()).toContain(`-${tagType}`)
    })

    test('renders with the "-disabled" when disabled prop is true', () => {
      const wrapper = mount(HaTag, {
        props: {
          disabled: true,
        },
      })

      expect(wrapper.classes()).toContain('-disabled')
    })

    test('does not render with the "-disabled" when disabled prop is false', () => {
      const wrapper = mount(HaTag, {
        props: {
          disabled: false,
        },
      })

      expect(wrapper.classes()).not.toContain('-disabled')
    })

    test('renders with the "-clickable" when clickable prop is true', () => {
      const wrapper = mount(HaTag, {
        props: {
          clickable: true,
        },
      })

      expect(wrapper.classes()).toContain('-clickable')
    })

    test('does not render with the "-clickable" when clickable prop is false', () => {
      const wrapper = mount(HaTag, {
        props: {
          clickable: false,
        },
      })

      expect(wrapper.classes()).not.toContain('-clickable')
    })
  })

  describe('emit', () => {
    test('handles click event when clickable and not disabled', async () => {
      const wrapper = mount(HaTag, {
        props: {
          disabled: false,
          clickable: true,
        },
      })

      await wrapper.trigger('click')
      expect(wrapper.emitted('click')).toHaveLength(1)
      expect(wrapper.emitted('click')).toBeTruthy()
    })

    test('does not handle click event when disabled', async () => {
      const wrapper = mount(HaTag, {
        props: {
          disabled: true,
          clickable: true,
        },
      })

      await wrapper.trigger('click')
      expect(wrapper.emitted('click')).toBeFalsy()
    })
  })
})
