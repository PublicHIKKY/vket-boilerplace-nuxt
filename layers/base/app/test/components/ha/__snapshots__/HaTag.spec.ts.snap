// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HaTag > mount component 1`] = `"<span data-v-c08b7397="" class="ha-tag -primary" disabled="false"></span>"`;
